packager/mux/mux.go

```diff
@@ -37,82 +37,134 @@ import (
 	"net/http"
 	"net/url"
 	"strings"
-
-	"github.com/ampproject/amppackager/packager/util"
 )
 
+// routingRule maps a URL path prefix to three entities:
+// * suffixValidatorFunc - a function that validates the suffix of URL path,
+// * handler - an http.Handler that should handle such prefix,
+// * handlerPrometheusAlias - the alias for reporting Prometheus metrics.
+type routingRule struct {
+	URLPathPrefix          string
+	suffixValidatorFunc    func(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int)
+	handler                http.Handler
+	handlerPrometheusAlias string
+}
+
+// mux stores a routingMatrix, an array of routing rules that define the mux'
+// routing logic. Note that the order of rules in the matrix is important: the
+// first matching rule will be applied by ServeHTTP, so the rule for a
+// particular prefix should precede the rule for it's sub-prefix. E.g. the rule
+// for routing requests prefixed with “/priv/doc/” (w/ trailing slash) must go
+// before the rule for routing requests prefixed with “/priv/doc” (w/o trailing
+// slash).
+// The last rule in the matrix must have empty URLPathPrefix, so it would match
+// any URL. This ensures there's at least one matching rule for any URL.
 type mux struct {
-	certCache   http.Handler
-	signer      http.Handler
-	validityMap http.Handler
-	healthz     http.Handler
+	routingMatrix []routingRule
+}
+
+// return404 is a URL Path Suffix Validator that always returns 404.
+func return404(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	*errorMsg, *errorCode = "404 page not found", http.StatusNotFound
+}
+
+// expectNoSuffix is a URL Path Suffix Validator that expects an empty suffix.
+func expectNoSuffix(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	if suffix != "" {
+		return404(suffix, req, params, errorMsg, errorCode)
+	}
+}
+
+// expectSignerQuery is a URL Path Suffix Validator specific to signer requests.
+func expectSignerQuery(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	(*params)["signURL"] = suffix
+	if req.URL.RawQuery != "" {
+		(*params)["signURL"] += "?" + req.URL.RawQuery
+	}
+}
+
+// expectCertQuery is a URL Path Suffix Validator specific to cert requests.
+func expectCertQuery(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	unescaped, err := url.PathUnescape(suffix)
+	if err != nil {
+		*errorMsg, *errorCode = "400 bad request - bad URL encoding", http.StatusBadRequest
+	} else {
+		(*params)["certName"] = unescaped
+	}
```

Comment:
Thanks for keeping the logic of these so close to the original, so the refactoring is easy to follow!